libs/wingc/src/lsp/rename_visitor.rs

```diff
@@ -154,10 +154,49 @@ impl<'a> Visit<'a> for RenameVisitor<'a> {
 	}
 
 	fn visit_symbol(&mut self, node: &'a Symbol) {
-		self.add_reference_symbol(node);
+		self.add_reference_symbol(node, None);
 		crate::visit::visit_symbol(self, node);
 	}
 
+	fn visit_expr(&mut self, node: &'a Expr) {
+		if let ExprKind::JsonMapLiteral { fields, .. } = &node.kind {
+			let type_ = self.types.maybe_unwrap_inference(self.types.get_expr_type(node));
+			let type_ = *if let Some(type_) = self.types.get_type_from_json_cast(node.id) {
+				*type_
+			} else {
+				type_
+			}
+			.maybe_unwrap_option();
+
+			if let Some(c) = type_.as_struct() {
+				for (field, ..) in fields {
+					self.add_reference_symbol(field, Some(&UnsafeRef::from(&c.env)));
+				}
+			}
+		}
+		crate::visit::visit_expr(self, node);
+	}
+
+	fn visit_reference(&mut self, node: &'a Reference) {
+		match node {
+			Reference::InstanceMember { property, object, .. } => {
+				let object_type = self.types.get_expr_id_type_ref(object.id);
+				if let Some(expr_type) = object_type.as_class() {
+					self.add_reference_symbol(property, Some(&UnsafeRef::from(&expr_type.env)));
+				}
+				if let Some(expr_type) = object_type.as_struct() {
+					self.add_reference_symbol(property, Some(&UnsafeRef::from(&expr_type.env)));
+				}
+				if let Some(expr_type) = object_type.as_interface() {
+					self.add_reference_symbol(property, Some(&UnsafeRef::from(&expr_type.env)));
+				}
```

Comment:
I think as_env() can be used here to simplify things

```suggestion
				if let Some(inner_env) = object_type.as_env() {
					self.add_reference_symbol(property, Some(inner_env));
				}
```

Or even this, if the logic is the same

```suggestion
				self.add_reference_symbol(property, object_type.as_env());
```

Reply:
That's a good one! added :)